test/libsolidity/SemanticTest.cpp

```diff
@@ -239,12 +237,51 @@ TestCase::TestResult SemanticTest::run(ostream& _stream, string const& _linePref
 	return TestResult::Success;
 }
 
-void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool) const
+void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool _formatted) const
 {
-	stringstream stream(m_source);
-	string line;
-	while (getline(stream, line))
-		_stream << _linePrefix << line << endl;
+	if (m_sources.sources.empty())
+		return;
+
+	bool outputSourceNames = true;
+	if (m_sources.sources.size() == 1 && m_sources.sources.begin()->first.empty())
+		outputSourceNames = false;
+
+	if (_formatted)
+		for (auto const& [name, source]: m_sources.sources)
+		{
+			if (source.empty())
+				continue;
+
+			if (outputSourceNames)
+				_stream << _linePrefix << formatting::CYAN << "==== Source: " << name << " ====" << formatting::RESET << endl;
+			vector<char const*> sourceFormatting(source.length(), formatting::RESET);
+
+			_stream << _linePrefix << sourceFormatting.front() << source.front();
+			for (size_t i = 1; i < source.length(); i++)
+			{
+				if (sourceFormatting[i] != sourceFormatting[i - 1])
+					_stream << sourceFormatting[i];
+				if (source[i] != '\n')
+					_stream << source[i];
+				else
+				{
+					_stream << formatting::RESET << endl;
+					if (i + 1 < source.length())
+						_stream << _linePrefix << sourceFormatting[i];
+				}
+			}
+			_stream << formatting::RESET;
+		}
+	else
+		for (auto const& [name, source]: m_sources.sources)
+		{
+			if (outputSourceNames)
+				_stream << _linePrefix << "==== Source: " + name << " ====" << endl;
+			stringstream stream(source);
+			string line;
+			while (getline(stream, line))
+				_stream << _linePrefix << line << endl;
+		}
```

Comment:
This is identical to https://github.com/ethereum/solidity/blob/e419e904d3ce2625995175dc155b911d2b4d2340/test/CommonSyntaxTest.cpp#L92

with the following lines (error parsing) removed

https://github.com/ethereum/solidity/blob/e419e904d3ce2625995175dc155b911d2b4d2340/test/CommonSyntaxTest.cpp#L110-L124

I didn't introduce error parsing with this PR because legacy code did not either. If we would like to support syntax error diagnosis for semantic tests, we should consider doing it here as well. But perhaps that is another issue.

Reply:
I think it would make sense to use the same function for both.

At the same time, I would like to merge this rather sooner than later because we want to add some tests, so maybe it can be done in its own PR.

Reply:
Okay, let's move the refactoring to a future PR then.

Reply:
May be that PR can also make error highlighting uniform across syntactic and semantic tests.

Reply:
created https://github.com/ethereum/solidity/issues/9250 to track this